### PR TITLE
fix: fix accumulation of loss across microbatches

### LINKED ISSUE #212: Correctly handle loss normalization over microbatches
**Describe the bug**

Currently, we [divide the loss by the number of microbatches](https://github.com/NVIDIA/reinforcer/blob/5ff10f61347c2d407ea419e13c24f85f5e23b0b3/nemo_reinforcer/models/policy/hf_policy.py#L336) before backpropagating. However, if a microbatch has no valid samples (e.g. we're running with a microbatch size of 1 and the sample is masked because it exceeds max_seq_length), this normalization constant is incorrect. We should instead normalize by the number of _valid_ microbatches per batch

### LINKED ISSUE #213: Fix gradient accumulation bug in SFT
**Describe the bug**

Bug is described at https://huggingface.co/blog/gradient_accumulation

**Steps/Code to reproduce bug**

Please list *minimal* steps or code snippet for us to be able to reproduce the bug.

A helpful guide on on how to craft a minimal bug report  http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports. 

**Expected behavior**

**Environment overview (please complete the following information)**

 - Environment location: [Bare-metal, Docker, Cloud(specify cloud provider - AWS, Azure, GCP, Collab)]
 - Method of install: [pip install or from source]. Please specify exact commands you used to install.
 - If method of install is [Docker], provide `docker pull` & `docker run` commands used

**Environment details**

If NVIDIA docker image is used you don't need to specify these.
Otherwise, please provide:
- OS version
- PyTorch version
- Python version

**Additional context**

Example: GPU model

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

**Add a one line overview of what this PR aims to accomplish.**

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

closes #212 
closes #213 

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA/reinforcer/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA/reinforcer/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA/reinforcer/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: nemo_reinforcer/algorithms/loss_functions.py [hunk 1]
```diff
@@ -234,19 +254,13 @@ def __call__(
...
-            num_unmasked_tokens = num_unmasked_tokens.item()
+            ## scale by the total number of tokens in the batch
+            loss = -masked_global_mean(token_logprobs, mask, normalization_factor)
 
         return loss, {
             "loss": loss.item() if loss.ndim == 0 else loss,
-            "num_unmasked_tokens": num_unmasked_tokens,
-            "total_tokens": mask.numel(),
+            "total_tokens_per_mb": mask.numel(),
+            "num_unmasked_tokens": normalization_factor.item(),
```
Comment (terrykong) [inline nemo_reinforcer/algorithms/loss_functions.py]: should this be computed in a way that doesn't rely on normalization_factor? this renaming may be surprising if used in a different context

File: nemo_reinforcer/algorithms/dpo.py [hunk 2]
```diff
@@ -303,7 +303,10 @@ def validate(
 
             else:
                 for k, v in val_results["all_mb_metrics"].items():
-                    val_metrics[k] += np.mean(v).item()
+                    if k in {"loss", "num_valid_samples", "num_unmasked_tokens"}:
```
Comment (terrykong) [inline nemo_reinforcer/algorithms/dpo.py]: is it possible to formalize the reduction (sum vs. mean) somewhere closer to the loss for readability and make it easier to update this set? someone needs to now look at the loss function and then just into the algorithm to see whether avg or sum

File: nemo_reinforcer/algorithms/utils.py [hunk 3]
```diff
@@ -208,3 +201,8 @@ def get_tokenizer(tokenizer_config: TokenizerConfig) -> AutoTokenizer:
         print("No chat template provided, using tokenizer's default")
 
     return tokenizer
+
+
+def masked_global_mean(values, mask, normalization_factor):
```
Comment (terrykong) [inline nemo_reinforcer/algorithms/utils.py]: do we get those pytorch warnings anymore?

File: nemo_reinforcer/models/policy/dtensor_policy_worker.py [hunk 4]
```diff
@@ -289,6 +290,31 @@ def train(
...
+                    ## get number of tokens in the global batch
+                    normalization_factor = torch.sum(
+                        data["token_mask"][:, 1:] * data["sample_mask"].unsqueeze(-1)
+                    )
+                    torch.distributed.all_reduce(
+                        normalization_factor, group=self.dp_mesh.get_group()
+                    )
+                elif loss_fn.loss_type == LossType.SAMPLE_LEVEL:
+                    ## get number of valid samples in the global batch
+                    assert "sample_mask" in data, (
```
Comment (terrykong) [inline nemo_reinforcer/models/policy/dtensor_policy_worker.py]: assert is already done above

File: nemo_reinforcer/models/policy/fsdp1_policy_worker.py [hunk 5]
```diff
@@ -258,6 +259,27 @@ def train(
...
+                        "token_mask must be present in the data when using token-level loss"
+                    )
+                    ## get number of tokens in the global batch
+                    normalization_factor = torch.sum(
+                        data["token_mask"][:, 1:] * data["sample_mask"].unsqueeze(-1)
+                    )
+                    torch.distributed.all_reduce(normalization_factor)
+                elif loss_fn.loss_type == LossType.SAMPLE_LEVEL:
+                    ## get number of valid samples in the global batch
+                    assert "sample_mask" in data, (
```
Comment (terrykong) [inline nemo_reinforcer/models/policy/fsdp1_policy_worker.py]: same comment about assert

File: nemo_reinforcer/algorithms/loss_functions.py (hunk 1 above)
Comment (ashors1) [reply nemo_reinforcer/algorithms/loss_functions.py]: Just pushed a change that fixes this

File: nemo_reinforcer/algorithms/dpo.py (hunk 2 above)
Comment (ashors1) [reply nemo_reinforcer/algorithms/dpo.py]: Yeah, I also just fixed this. The new idea is that all metrics returned from the loss function should be summed across microbatches, and the onus is on the developer to make sure that's the case. Then we can just sum across all metrics and not worry about specific keys. There are a few exceptions where we want to take the mean (e.g. learning rate), but in general those are shared across algorithms so I think it's okay to handle those separately.

File: nemo_reinforcer/algorithms/utils.py (hunk 3 above)
Comment (ashors1) [reply nemo_reinforcer/algorithms/utils.py]: Could you share an example? I'm not seeing any warnings, but maybe I'm just not triggering them

File: nemo_reinforcer/models/policy/dtensor_policy_worker.py [hunk 6]
```diff
@@ -289,6 +290,28 @@ def train(
             losses = []
             all_mb_metrics = []
             for gb_start in range(0, dataset_size, local_gbs):
+                assert "sample_mask" in data, "sample_mask must be present in the data!"
```
Comment (SahilJain314) [inline nemo_reinforcer/models/policy/dtensor_policy_worker.py]: Let's also add a type annotation to the input data dict. i.e. BatchedDataDict[TrainDataSpec]

File: nemo_reinforcer/algorithms/loss_functions.py [hunk 7]
```diff
@@ -195,10 +212,13 @@ def __call__(
 class NLLLoss(LossFunction):
     """Negative Log Likelihood Loss function."""
 
+    loss_type = LossType.TOKEN_LEVEL
+
     def __call__(
         self,
         next_token_logits: torch.Tensor,
         data: BatchedDataDict,
+        normalization_factor: torch.Tensor,
```
Comment (SahilJain314) [inline nemo_reinforcer/algorithms/loss_functions.py]: normalization_factor -> num_valid_tokens_or_seqs?

File: nemo_reinforcer/algorithms/utils.py [hunk 8]
```diff
@@ -119,11 +119,9 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
-# need to surpress the masked tensor warnings from pytorch
-@surpress_user_warnings
-def masked_mean(values, mask, dim=None):
-    """Masks values with mask, and computes the mean of the values using the masked values."""
-    return (values * mask).sum(dim=dim) / (mask.sum(dim=dim) + 1e-8)
+def masked_global_mean(values, mask, normalization_factor):
```
Comment (SahilJain314) [inline nemo_reinforcer/algorithms/utils.py]: maybe masked_mean with an optional global_normalization_factor? (global to me implies there is a communication being done)

File: nemo_rl/algorithms/loss_functions.py [hunk 9]
```diff
@@ -102,7 +114,21 @@ def __call__(
...
+            ).item()
+        else:
+            # first average over tokens per sample, then average over samples
+            mult_prob_error = masked_mean(
+                masked_mean(torch.exp(lp_error), token_mask, dim=-1),
+                sample_mask,
+                global_normalization_factor=total_valid_tokens_or_seqs,
+            ).item()
+
         if mult_prob_error == 0.0:
```
Comment (SahilJain314) [inline nemo_rl/algorithms/loss_functions.py]: can we remove this hack now that we have a correct aggregation scheme?
